Honor “lang” attribute in no namespace only on HTML and SVG elements

sideshowbarker opened the pull request:
#### 6599021230e5957efed742285983012bd5b5703f
<pre>
Honor “lang” attribute in no namespace only on HTML and SVG elements
<a href="https://bugs.webkit.org/show_bug.cgi?id=263655">https://bugs.webkit.org/show_bug.cgi?id=263655</a>

Reviewed by Ryosuke Niwa.

<a href="https://github.com/whatwg/html/pull/9882/files">https://github.com/whatwg/html/pull/9882/files</a> added the phrase “or an
element in the SVG namespace” in the following HTML spec text:

&gt; To determine the language of a node, user agents must use the first
appropriate step in the following list:
&gt; …
&gt; - If the node is an HTML element or an element in the SVG namespace,
    and it has a lang in no namespace attribute set: Use the value of
    that attribute.

And that updated spec text is now part of the HTML standard, at
<a href="https://html.spec.whatwg.org/#the-lang-and-xml">https://html.spec.whatwg.org/#the-lang-and-xml</a>:lang-attributes:svg-namespace

Gecko has already had that behavior implemented for quite a while, and
Blink was recently updated to implement it as well.

So this change fully aligns WebKit with that updated spec text, and with
the current behavior in Gecko and Blink. However, this change does that
by updating the code to *restrict* the “lang” attribute in no namespace
to *only* being honored on HTML and SVG elements.

Otherwise, without this code change, WebKit honors the “lang” attribute
in no namespace regardless of what element it’s specified on (that is,
not just for HTML elements and SVG elements) — which breaks conformance
with the spec, and breaks interoperability with Gecko and Blink.

* LayoutTests/imported/w3c/web-platform-tests/html/dom/elements/global-attributes/lang-attribute.window-expected.txt:
* Source/WebCore/dom/Element.cpp:
(WebCore::Element::attributeChanged):

Canonical link: <a href="https://commits.webkit.org/273726@main">https://commits.webkit.org/273726@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/67f34695bd05992f99053e13c1ee086c5921cf2e

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/36409 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/15359 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/38627 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/39114 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/32706 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/37639 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/17811 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/51/builds/12392 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/31352 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/36970 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/12959 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/32261 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/11356 "Passed tests") | 
| | [  ~~🧪 ios-wk2-wpt~~](https://ews-build.webkit.org/#/builders/42/builds/11384 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/32494 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/40359 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/33026 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/32843 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/37322 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/11601 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/50/builds/9475 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/35422 "Passed tests") | 
| | [  ~~🛠 tv-sim~~](https://ews-build.webkit.org/#/builders/46/builds/13309 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/8264 "Built successfully and passed tests") | [  ~~🛠 watch~~](https://ews-build.webkit.org/#/builders/43/builds/12051 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/12468 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

annevk reviewed (commented):
Non-reviewer caveats apply: I think we should instead update `hasLangAttr()` to include these checks. Because now other things building on top of `hasLangAttr()` become incorrect of which you caught some, but not in the most elegant way (it seems more expensive to turn has checks into checks on returned strings) and it's also not clear to me you caught all as `hasLangAttr()` is still around and `hasLanguageAttribute()` is too.

Hope that helps!

Aside: looking at the code it seems we don't support shadow tree inheritance for language either. Maybe a fun follow-up issue?

annevk reviewed (commented):
The logic of this seems sound. Thanks!

Now you need an actual WebKit reviewer for the remainder. Maybe @cdumez or @rniwa?

annevk reviewed (commented):
Not actually having looked at benchmark numbers I suspect it would be better if we instead changed when we called `setHasLangAttr()`. In particular with selectors we'd do more lookups and so having that be a single bit check seems best.

And unfortunately this also needs rebasing due to Ahmad's recent work on importing that directory.

rniwa commented on Source/WebCore/dom/Element.cpp:
```diff
@@ -2141,7 +2141,7 @@ void Element::attributeChanged(const QualifiedName& name, const AtomString& oldV
     case AttributeNames::langAttr: {
         if (name == HTMLNames::langAttr)
-            setHasLangAttr(!newValue.isNull());
+            setHasLangAttr(!newValue.isNull() && ((isHTMLElement() || isSVGElement())));
```
Nit: You have an extra parenthesis here. Just do: `!newValue.isNull() && (isHTMLElement() || isSVGElement())`

sideshowbarker replied:
Thanks for catching that — fixed now

bnham committed:
Add support for updatable list of domains with restricted openers
https://bugs.webkit.org/show_bug.cgi?id=268185
rdar://problem/121677395

Reviewed by Wenson Hsieh.

This adds support for a list of domains that will receive restricted window opener access. The list
is maintained and updated through WebPrivacy.

I didn't add a method to override the list for testing purposes yet. I'll do that in a future patch
which actually uses these interfaces.

* Source/WebCore/PAL/pal/spi/cocoa/WebPrivacySPI.h:
* Source/WebKit/Platform/cocoa/WebPrivacyHelpers.h:
* Source/WebKit/Platform/cocoa/WebPrivacyHelpers.mm:
(-[WKWebPrivacyNotificationListener listenForRestrictedOpenerDomainsChanges:]):
(-[WKWebPrivacyNotificationListener didUpdate:]):
(WebKit::RestrictedOpenerDomainsController::shared):
(WebKit::RestrictedOpenerDomainsController::RestrictedOpenerDomainsController):
(WebKit::restrictedOpenerType):
(WebKit::RestrictedOpenerDomainsController::populate):
(WebKit::RestrictedOpenerDomainsController::lookup const):
* Source/WebKit/WebKit.xcodeproj/project.pbxproj:

Canonical link: https://commits.webkit.org/273693@main

charliewolfe committed:
REGRESSION (273517@main): [iOS17] 2 tests in platform/ios/fast/scrolling are flaky failures
https://bugs.webkit.org/show_bug.cgi?id=268333
rdar://121882801

Reviewed by Aditya Keerthi.

Wait for a presentation update before completing the test.

* LayoutTests/platform/ios/fast/scrolling/find-text-in-overflow-node-indicator-position-limit.html:
* LayoutTests/platform/ios/fast/scrolling/find-text-in-overflow-node-indicator-position.html:

Canonical link: https://commits.webkit.org/273694@main

aestes committed:
[Cocoa] Create a top-level folder in WebKit for WebKitSwift files
https://bugs.webkit.org/show_bug.cgi?id=268301
rdar://121858909

Reviewed by Ben Nham.

Moved GroupSession.swift and MarketplaceKitWrapper.swift from WebKit/UIProcess/Cocoa to WebKit/WebKitSwift.
Sorted the top-level file and folder references in the WebKit project.

* Source/WebKit/WebKit.xcodeproj/project.pbxproj:
* Source/WebKit/WebKitSwift/GroupActivities/GroupSession.swift: Renamed from Source/WebKit/UIProcess/Cocoa/GroupActivities/WKGroupSession.swift.
* Source/WebKit/WebKitSwift/MarketplaceKit/MarketplaceKitWrapper.swift: Renamed from Source/WebKit/UIProcess/Cocoa/MarketplaceKitWrapper.swift.

Canonical link: https://commits.webkit.org/273695@main

Commit:
WebCore::CAAudioStreamDescription::m_platformDescription::type is uninitialized in constructor
https://bugs.webkit.org/show_bug.cgi?id=268255
<rdar://121809321>

Reviewed by Eric Carlson.

* Source/WebCore/platform/audio/AudioStreamDescription.h:
(WebCore::PlatformDescription::type):
- Set default value to `None` enum.

Canonical link: https://commits.webkit.org/273696@main

annevk committed:
HTML parser: execute self-closing SVG script element
https://bugs.webkit.org/show_bug.cgi?id=268168

Reviewed by Ryosuke Niwa and Chris Dumez.

The code is largely structured like the HTML Standard, except we
currently do not fake an end tag as needed for m_scriptToProcess to be
properly assigned to.

As foreign content supports self-closing start tags we also have to
pretend the start tag is a non-self-closing one. This results in some
slight code smell, but it seems acceptable enough.

Gecko does this correctly already.

Aligns WPT with this commit:
https://github.com/web-platform-tests/wpt/commit/eef173b5b9ee626ff9e144c80f8082af2f086e4c

* LayoutTests/imported/w3c/web-platform-tests/html/syntax/parsing/support/svg-script-self-closing.js: Added.
* LayoutTests/imported/w3c/web-platform-tests/html/syntax/parsing/support/w3c-import.log:
* LayoutTests/imported/w3c/web-platform-tests/html/syntax/parsing/unclosed-svg-script-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/html/syntax/parsing/unclosed-svg-script.html:
* Source/WebCore/html/parser/AtomHTMLToken.h:
(WebCore::AtomHTMLToken::setSelfClosingToFalse):
* Source/WebCore/html/parser/HTMLTreeBuilder.cpp:
(WebCore::HTMLTreeBuilder::processTokenInForeignContent):

Canonical link: https://commits.webkit.org/273697@main

gsnedders committed:
Drop support for Lighttpd
https://bugs.webkit.org/show_bug.cgi?id=268292

Reviewed by Alexey Proskuryakov.

Every port nowadays has `_use_apache` returning True, so this is now all
dead code.

* Tools/Scripts/webkitpy/layout_tests/servers/http_server.py: Removed.
* Tools/Scripts/webkitpy/layout_tests/servers/http_server_base.py:
* Tools/Scripts/webkitpy/layout_tests/servers/lighttpd.conf: Removed.
* Tools/Scripts/webkitpy/layout_tests/servers/websocket_server.py:
(PyWebSocket):
(PyWebSocket.__init__):
(PyWebSocket._stop_running_server):
(PyWebSocket._check_and_kill):
(PyWebSocket._is_server_running_on_all_ports):
* Tools/Scripts/webkitpy/layout_tests/servers/websocket_server_unittest.py: Renamed from Tools/Scripts/webkitpy/layout_tests/servers/http_server_unittest.py.
(TestWebsocketServer):
(TestWebsocketServer.test_start_cmd):
(TestWebsocketServer.test_win32_start_and_stop):
(TestWebsocketServer.test_win32_start_and_stop.wait_for_action):
(TestWebsocketServer.test_win32_start_and_stop.mock_returns):
(TestWebsocketServer.test_win32_start_and_stop.mock_returns.return_value_thunk):
* Tools/Scripts/webkitpy/port/base.py:
(Port.check_httpd):
(Port.start_http_server):
(Port):
(Port.path_to_api_test_binaries):
(Port._uses_apache): Deleted.
(Port._path_to_lighttpd): Deleted.
(Port._path_to_lighttpd_modules): Deleted.
* Tools/Scripts/webkitpy/port/test.py:
* Tools/Scripts/webkitpy/port/win.py:
(WinPort):
(WinPort._path_to_apache):
(WinPort._uses_apache): Deleted.
(WinPort._path_to_lighttpd): Deleted.
(WinPort._path_to_lighttpd_modules): Deleted.
* Tools/Scripts/webkitpy/test/main.py:
(main):

Canonical link: https://commits.webkit.org/273698@main

hortont424 committed:
[UnifiedPDF] Make all painting be in-process
https://bugs.webkit.org/show_bug.cgi?id=268325
<rdar://problem/121385041>

Reviewed by Simon Fraser.

UnifiedPDFPlugin currently paints PDF content into an ImageBuffer in the
Web Content process (to keep PDF out of the GPU process, for security),
and then paints that ImageBuffer into the tiles in the GPU process (to follow
the normal flow of painting).

Skip this two step process by making the layers for UnifiedPDFPlugin
have in-process software backing store and painting directly into them.

* Source/WebCore/platform/graphics/GraphicsLayerClient.h:
(WebCore::GraphicsLayerClient::layerNeedsPlatformContext const):
* Source/WebCore/platform/graphics/ca/GraphicsLayerCA.cpp:
(WebCore::GraphicsLayerCA::platformCALayerNeedsPlatformContext const):
* Source/WebCore/platform/graphics/ca/GraphicsLayerCA.h:
* Source/WebCore/platform/graphics/ca/PlatformCALayer.h:
* Source/WebCore/platform/graphics/ca/PlatformCALayer.mm:
(WebCore::PlatformCALayer::needsPlatformContext const):
* Source/WebCore/platform/graphics/ca/PlatformCALayerClient.h:
(WebCore::PlatformCALayerClient::platformCALayerNeedsPlatformContext const):
* Source/WebCore/platform/graphics/ca/TileGrid.cpp:
(WebCore::TileGrid::platformCALayerNeedsPlatformContext const):
* Source/WebCore/platform/graphics/ca/TileGrid.h:
* Source/WebKit/Shared/RemoteLayerTree/RemoteLayerBackingStoreCollection.mm:
(WebKit::RemoteLayerBackingStoreCollection::createRemoteLayerBackingStore):
Plumb around a "layer needs a platform context" bit, and, if set,
force in-process rendering.

* Source/WebKit/WebProcess/Plugins/PDF/UnifiedPDF/UnifiedPDFPlugin.h:
Always say that UnifiedPDFPlugin needs in-process rendering.

* Source/WebKit/WebProcess/Plugins/PDF/UnifiedPDF/UnifiedPDFPlugin.mm:
(WebKit::UnifiedPDFPlugin::paintPDFContent):
Avoid painting into an extra bitmap here.

Canonical link: https://commits.webkit.org/273699@main

rcaliman-apple committed:
Web Inspector: Update COMPATIBILITY (macOS X.0, iOS X.0) comments with correct versions
https://bugs.webkit.org/show_bug.cgi?id=268176
rdar://121675559

Reviewed by Patrick Angle.

Retroactively update backwards compatibility comments that were
missing corresponding version numbers for when they started applying.

Related to changes introduced in
- https://commits.webkit.org/267488@main
- https://commits.webkit.org/264172@main (relying on protocol changes in https://commits.webkit.org/261512@main for the same version check)

* Source/WebInspectorUI/UserInterface/Controllers/CSSManager.js:
(WI.CSSManager.prototype._mainResourceDidChange):
* Source/WebInspectorUI/UserInterface/Models/Canvas.js:
(WI.Canvas.fromPayload):
(WI.Canvas.prototype.sizeChanged):

Canonical link: https://commits.webkit.org/273700@main

Ahmad-S792 committed:
Sync and update 'which' across KeyboardEvent, MouseEvent and UIEvent interfaces as per web specification

https://bugs.webkit.org/show_bug.cgi?id=253366
rdar://problem/106580687

Reviewed by Ryosuke Niwa.

This patch aligns WebKit with Gecko / Firefox, Blink / Chromium and Web Specifications [1], [2] and [3].

[1] https://w3c.github.io/uievents/#idl-uievent
[2] https://w3c.github.io/uievents/#idl-keyboardevent
[3] https://w3c.github.io/uievents/#idl-mouseevent

This patch modifies 'which' to be 'unsigned' as per web-specification and also remove it from 'KeyboardEvent'.

Merge: https://chromium.googlesource.com/chromium/src.git/+/cf13270c47f29b21ce7a8c937f598989cd66c840

* Source/WebCore/dom/KeyboardEvent.cpp:
(KeyboardEvent::which):
* Source/WebCore/dom/KeyboardEvent.h:
* Source/WebCore/dom/KeyboardEvent.idl:
* Source/WebCore/dom/MouseEvent.cpp:
(MouseEvent::which):
* Source/WebCore/dom/MouseEvent.h:
* Source/WebCore/dom/UIEvent.cpp:
(UIEvent::which):
* Source/WebCore/dom/UIEvent.h:
* Source/WebCore/dom/UIEvent.idl: Remove FIXME and align with web-spec

Canonical link: https://commits.webkit.org/273701@main

szewai committed:
Make CookieChangeObserver subclass CanMakeWeakPtr
https://bugs.webkit.org/show_bug.cgi?id=268106
rdar://121623707

Reviewed by Chris Dumez.

We should not use CheckedPtr for class data members, i.e. NetworkStorageSession should store WeakPtr of
CookieChangeObserver instead of CheckedPtr.

* Source/WebCore/platform/network/NetworkStorageSession.h:
* Source/WebCore/platform/network/cocoa/NetworkStorageSessionCocoa.mm:
(WebCore::NetworkStorageSession::registerCookieChangeListenersIfNecessary):
(WebCore::NetworkStorageSession::startListeningForCookieChangeNotifications):
(WebCore::NetworkStorageSession::stopListeningForCookieChangeNotifications):

Canonical link: https://commits.webkit.org/273702@main

xeenon committed:
Remove contains check for tabs and windows.
https://webkit.org/b/268342
rdar://problem/121887064

Reviewed by Brian Weinstein.

The "Window returned by windowForWebExtensionContext: does not contain the tab" exception is hard
to avoid when the tabs is being created, and not in a window yet. This exception is causing more
issues than is solves currently, so we should just remove it for now.

All existing code paths that use window() or index() properly check for nullptr or notFound.

Also adopt the fallbackAction() helper in a couple places I missed.

* Source/WebKit/UIProcess/Extensions/Cocoa/WebExtensionActionCocoa.mm:
(WebKit::WebExtensionAction::hasUnreadBadgeText const): Use fallbackAction().
(WebKit::WebExtensionAction::setHasUnreadBadgeText): Ditto.
* Source/WebKit/UIProcess/Extensions/Cocoa/WebExtensionContextCocoa.mm:
(WebKit::WebExtensionContext::didCloseTab): Stop using SkipContainsCheck::Yes.
* Source/WebKit/UIProcess/Extensions/Cocoa/WebExtensionTabCocoa.mm:
(WebKit::WebExtensionTab::window const): Removed SkipContainsCheck param.
(WebKit::WebExtensionTab::index const): Remove contains check.
(WebKit::WebExtensionTab::isPrivate const): Stop using SkipContainsCheck::Yes.
* Source/WebKit/UIProcess/Extensions/WebExtensionTab.h:

Canonical link: https://commits.webkit.org/273703@main

cdumez committed:
Fix typo in NetworkProcess::acquireLockedFileGrant()
https://bugs.webkit.org/show_bug.cgi?id=268327

Reviewed by Per Arne Vollan.

* Source/WebKit/NetworkProcess/NetworkProcess.cpp:
(WebKit::NetworkProcess::setIsHoldingLockedFiles):
* Source/WebKit/NetworkProcess/NetworkProcess.h:
* Source/WebKit/NetworkProcess/cocoa/NetworkProcessCocoa.mm:
(WebKit::NetworkProcess::acquireLockedFileGrant):
(WebKit::NetworkProcess::aqcuireLockedFileGrant): Deleted.

Canonical link: https://commits.webkit.org/273704@main

rr-codes committed:
Unreviewed build fix.

Fix the build after 273687@main.

* Source/WebCore/dom/DocumentMarkerController.cpp:
(WebCore::DocumentMarkerController::showMarkers const):

Canonical link: https://commits.webkit.org/273705@main

martadarbinyan committed:
[Gardening]: REGRESSION (273601@main?): [ macOS Debug  ] ASSERTION FAILED:!m_inRemovedLastRefFunction result of 2 tests in inspector/canvas to be a flaky crash.
https://bugs.webkit.org/show_bug.cgi?id=268329
rdar://121879996

Unreviewed test gardening.

Adding test expectation

* LayoutTests/platform/mac-wk2/TestExpectations:

Canonical link: https://commits.webkit.org/273706@main

cdumez committed:
Crash under VideoPresentationManagerProxy::createLayerHostViewWithID()
https://bugs.webkit.org/show_bug.cgi?id=268320
rdar://121747495

Reviewed by Per Arne Vollan.

Null check the GPUProcess before dereferencing it. Even though we had a
GPUProcess to generate the id in the first place, the ID comes from IPC
from the WebProcess and things are just racing. We shouldn't assume the
GPUProcess is still running by the time we get the IPC.

* Source/WebKit/UIProcess/Cocoa/VideoPresentationManagerProxy.mm:
(WebKit::VideoPresentationManagerProxy::createLayerHostViewWithID):

Canonical link: https://commits.webkit.org/273707@main

cdumez committed:
Port std::monostate to the new IPC serialization format
https://bugs.webkit.org/show_bug.cgi?id=268312

Reviewed by Alex Christensen.

* Source/WebKit/Platform/IPC/ArgumentCoders.h:
(IPC::ArgumentCoder<std::monostate>::encode): Deleted.
(IPC::ArgumentCoder<std::monostate>::decode): Deleted.
* Source/WebKit/Shared/WTFArgumentCoders.serialization.in:

Canonical link: https://commits.webkit.org/273708@main

annevk committed:
More dynamicDowncast<> adoption in editing and OffscreenCanvas
https://bugs.webkit.org/show_bug.cgi?id=268256

Reviewed by Chris Dumez.

And increase RefPtr usage while there.

* Source/WebCore/editing/InsertLineBreakCommand.cpp:
(WebCore::InsertLineBreakCommand::doApply):
* Source/WebCore/editing/InsertParagraphSeparatorCommand.cpp:
(WebCore::isPhrasingContent):
(WebCore::InsertParagraphSeparatorCommand::doApply):
* Source/WebCore/editing/VisiblePosition.cpp:
(WebCore::midpoint):
* Source/WebCore/editing/cocoa/DataDetection.mm:
(WebCore::DataDetection::isDataDetectorLink):
(WebCore::searchForLinkRemovingExistingDDLinks):
(WebCore::DataDetection::detectContentInRange):
* Source/WebCore/html/OffscreenCanvas.cpp:
(WebCore::OffscreenCanvas::getContext):

Canonical link: https://commits.webkit.org/273709@main

nt1m committed:
REGRESSION(267658@main): `font-variant-emoji` values should not parse in `font-variant` shorthand when disabled
https://bugs.webkit.org/show_bug.cgi?id=268348
rdar://121890028

Reviewed by Matthieu Dubet.

`font-variant: emoji` or `font-variant: unicode` should not parse when `font-variant-emoji` is disabled.

* LayoutTests/fast/css/font-variant-emoji-disabled-expected.txt: Added.
* LayoutTests/fast/css/font-variant-emoji-disabled.html: Added.
* Source/WebCore/css/parser/CSSPropertyParser.cpp:
(WebCore::CSSPropertyParser::consumeFontVariantShorthand):

Canonical link: https://commits.webkit.org/273710@main

foopoiuyt committed:
[PlayStation] Build fix for missing optional after 273655@main
https://bugs.webkit.org/show_bug.cgi?id=268336

Unreviewed build fix.

Add the optional include to get std::optional in WebFrameLoaderClient.h

* Source/WebKit/WebProcess/WebCoreSupport/WebFrameLoaderClient.h:

Canonical link: https://commits.webkit.org/273711@main

pxlcoder committed:
Fix the build after 273687@main
https://bugs.webkit.org/show_bug.cgi?id=268352

Unreviewed build fix.

* Source/WebCore/dom/DocumentMarkerController.cpp:
(WebCore::DocumentMarkerController::showMarkers const):

Canonical link: https://commits.webkit.org/273712@main

jyavenard committed:
Add MessageSender::sendWithPromisedReply
https://bugs.webkit.org/show_bug.cgi?id=268032
rdar://121551204

Reviewed by Youenn Fablet.

Add sendWithPromisedReply methods to MessageSender for consistency with Connection.

* Source/WebKit/Platform/IPC/MessageSender.h:
* Source/WebKit/Platform/IPC/MessageSenderInlines.h:
(IPC::MessageSender::sendWithPromisedReply):
* Tools/TestWebKitAPI/Tests/IPC/MessageSenderTests.cpp:
(TestWebKitAPI::TEST_P):

Canonical link: https://commits.webkit.org/273713@main

alanbaradlay committed:
[IFC][Subpixel] InlineLevelBox::preferredLineHeight should return non-floored value
https://bugs.webkit.org/show_bug.cgi?id=268270

Reviewed by Antti Koivisto.

This integral-floor was added to match legacy line layout.

* Source/WebCore/layout/formattingContexts/inline/InlineLevelBox.h:
(WebCore::Layout::InlineLevelBox::preferredLineHeight const):

(/fast/forms/textAreaLineHeight-simple-lines.html ref test was added to check if these 2 codepaths were producing the same line height)

Canonical link: https://commits.webkit.org/273714@main

whsieh committed:
[iOS] Swiping back in an HTML Note with the keyboard presented causes SpringBoard to crash
https://bugs.webkit.org/show_bug.cgi?id=268334
rdar://121139498

Reviewed by Aditya Keerthi.

As a part of `BETextInput` adoption, we stopped relying on support from various private protocols
including `UITextInputMultiDocument`, which allows UIKit to `-_preserveFocusWithToken:destructively:`
and `-_restoreFocusWithToken:`. This was previously used to keep DOM elements focused while the web
view lost first responder status, during credential AutoFill, if the user searches for other
passwords. However, now that AutoFill UI in web views in third party apps is entirely managed by
`SafariViewService`, we no longer lose first responder anyways, so we don't need to adopt this SPI.

However, there is a corner case in Notes where, when performing a swipe navigation controller
transition while the web view is first responder and `-_editable` (in the WebKit SPI sense), we
relied on this SPI to keep the web view focused even as it temporarily loses first responder status
during the swipe transition. In iOS 17.4, we no longer preserve focus, which means that we'll end up
dispatching a blur event on the body.

This, in turn, causes Notes' injected JavaScript to post a message back to the UI process telling it
to make the web view non-editable. However, while this is happening, UIKit tells the web view to
become first responder again during the swipe transition, which (after an activity state update)
causes Notes to make the web view `-_editable` once again, underneath `WKInputDelegate` methods.
This endless loop of:

(UI Process)    Web view becomes editable, dispatching a message to the web process.
(Web Process)   Body is blurred, posting a message to the UI process.
…
(UI Process)    Web view receives posted message, and makes the web view non-editable.
(Web Process)   Body is focused as a result of enabling full-page editability, starting an input
                session and sending ElementDidFocus to the UI process.

...causes the keyboard to be presented and dismissed back-to-back so frequently, that SpringBoard
itself eventually crashes due to excessive memory use.

Mitigate this by restoring pre-iOS 17.4 behavior and conforming to `UITextInputMultiDocument`, but
only for WebKit SPI clients that set the `-_editable` flag on `WKWebView`. Notably, this excludes
Safari.

* Source/WebKit/UIProcess/ios/WKContentViewInteraction.mm:
(-[WKContentView _didChangeWebViewEditability]):

Canonical link: https://commits.webkit.org/273715@main

minorninth committed:
AX: should be able to compute AXPosition using the isolated tree
https://bugs.webkit.org/show_bug.cgi?id=267957
rdar://problem/121472048

Reviewed by Andres Gonzalez.

The relative frame that we're caching in the isolated tree is already
in the correct screen coordinates, all that's necessary is to cache
the screen relative position of the root, and then add them together.

We can update the cached root position when AXGeometryManager
updates object regions.

There isn't an easy way to write a layout test for this, as there
doesn't seem to be a way to move the window's origin.

In order to manually test this, I temporarily disabled AXRelativeFrame,
forcing VoiceOver to use AXPosition for every element, then I checked
that the bounds are correct when navigating and when moving and resizing
the window.

The main performance win I observe is when VoiceOver tries to focus on the
root web area when the page is busy. Without this patch, VoiceOver can
get blocked on the main thread, but with this patch it's no longer blocked.

* Source/WebCore/accessibility/AXGeometryManager.cpp:
(WebCore::AXGeometryManager::willUpdateObjectRegions):
* Source/WebCore/accessibility/isolatedtree/AXIsolatedObject.cpp:
(WebCore::AXIsolatedObject::screenRelativePosition const):
* Source/WebCore/accessibility/isolatedtree/AXIsolatedTree.cpp:
(WebCore::AXIsolatedTree::updateNodeProperties):
(WebCore::AXIsolatedTree::updateRootScreenRelativePosition):
* Source/WebCore/accessibility/isolatedtree/AXIsolatedTree.h:

Canonical link: https://commits.webkit.org/273716@main

twilco committed:
AX: Don't log AXStreamOptions::IdentifierAttribute if the id is empty
https://bugs.webkit.org/show_bug.cgi?id=268317
rdar://problem/121873562

Reviewed by Chris Fleizach.

Most objects have an empty id, so logging it for everything produces a
lot of noise. Also shorten parentObject to parentID, which is nice because
it's more symmetrical with the objectID property also dumped in this function.

* Source/WebCore/accessibility/AXLogger.cpp:
(WebCore::streamAXCoreObject):

Canonical link: https://commits.webkit.org/273717@main

rkirsling committed:
Fix PlayStation build following 273690@main
https://bugs.webkit.org/show_bug.cgi?id=268353

Unreviewed build fix.

* Source/WebCore/Modules/credentialmanagement/DigitalCredential.h:
Add missing include.

Canonical link: https://commits.webkit.org/273718@main

brentfulgham committed:
Protect WebProcess::unblockServicesRequiredByAccessibility from attempting to consume nullptr sandbox extensions
https://bugs.webkit.org/show_bug.cgi?id=268314
<rdar://121746329>

Reviewed by Per Arne Vollan.

The code in WebProcess::unblockServicesRequiredByAccessibility assumes that it
will always work with non-nullptr Sandbox Extensions. However, it's possible
for the SandboxExtension::create() method to return nullptr, so we should harden
this code to avoid crashing under such conditions.

* Source/WebKit/WebProcess/cocoa/WebProcessCocoa.mm:
(WebKit::WebProcess::unblockServicesRequiredByAccessibility):

Canonical link: https://commits.webkit.org/273719@main

rniwa committed:
Coalesce ContainerNode::childrenChanged in fast parser
https://bugs.webkit.org/show_bug.cgi?id=268324

Reviewed by Yusuke Suzuki.

This PR adds a new member function parserAppendChildIntoIsolatedTree to ContainerNode which inserts
child nodes without calling childrenChanged on the parent, and deploys it in HTMLFastPathParser.
childrenChanged gets called later in Element::finishParsingChildren via ContainerNode's
parserNotifyChildrenChanged. Newly introduced StateFlag::HeldBackChildrenChanged keeps track of
whether we held back childrenChanged calls when inserting child nodes or not.

Delaying calls to childrenChanged this way is safe in HTMLFastPathParser because the parsed tree
isn't exposed to scripts until the parsing ends so there won't be any event listeners that run
synchronously in response to DOM mutations.

Inspired by https://github.com/chromium/chromium/commit/b366b305efb5289f3bcf3bbbe549cbad82fc0daf

* Source/WebCore/dom/ContainerNode.cpp:
(WebCore::executeParserNodeInsertionIntoIsolatedTreeWithoutNotifyingParent): Added.
(WebCore::ContainerNode::parserAppendChildIntoIsolatedTree): Added.
(WebCore::ContainerNode::parserNotifyChildrenChanged): Added.
* Source/WebCore/dom/ContainerNode.h:
* Source/WebCore/dom/Document.cpp:
(WebCore::Document::setMarkupUnsafe):
* Source/WebCore/dom/Element.cpp:
(WebCore::Element::finishParsingChildren):
* Source/WebCore/dom/Node.h:
(WebCore::Node::hasHeldBackChildrenChanged const): Added.
(WebCore::Node::setHasHeldBackChildrenChanged): Added.
(WebCore::Node::clearHasHeldBackChildrenChanged): Added.
* Source/WebCore/html/parser/HTMLDocumentParserFastPath.cpp:
(WebCore::HTMLFastPathParser::parseChildren): Uses parserAppendChildIntoIsolatedTree now.
(WebCore::HTMLFastPathParser::parseContainerElement): Ditto.
(WebCore::HTMLFastPathParser::parseVoidElement): Ditto.

Canonical link: https://commits.webkit.org/273720@main

rniwa committed:
Introduce & deploy insertChildrenBeforeWithoutPreInsertionValidityCheck
https://bugs.webkit.org/show_bug.cgi?id=268206

Reviewed by Wenson Hsieh.

This PR refactors out ensurePreInsertionValidityForPhantomDocumentFragment and
insertChildrenBeforeWithoutPreInsertionValidityCheck and deploy them in ContainerNode's
append, prepend, and replaceChildren functions, and Node's before and after functions.

In addition, this PR adds two new tests for bugs that were caught incidentally by some
other tests during the development.

* LayoutTests/fast/dom/ChildNode-before-self-expected.txt: Added.
* LayoutTests/fast/dom/ChildNode-before-self.html: Added.
* LayoutTests/fast/dom/ParentNode-append-fragments-expected.html: Added.
* LayoutTests/fast/dom/ParentNode-append-fragments.html: Added.
* Source/WebCore/dom/ContainerNode.cpp:
(WebCore::ContainerNode::ensurePreInsertionValidityForPhantomDocumentFragment):
(WebCore::ContainerNode::insertChildrenBeforeWithoutPreInsertionValidityCheck):
(WebCore::ContainerNode::append):
(WebCore::ContainerNode::prepend):
(WebCore::ContainerNode::replaceChildren):
* Source/WebCore/dom/ContainerNode.h:
* Source/WebCore/dom/Node.cpp:
(WebCore::Node::convertNodesOrStringsIntoNodeVector):
(WebCore::Node::before):
(WebCore::Node::after):
* Source/WebCore/dom/Node.h:

Canonical link: https://commits.webkit.org/273721@main

rniwa committed:
Assertion failure in Document::ensureQuirks
https://bugs.webkit.org/show_bug.cgi?id=268344

Reviewed by Wenson Hsieh.

Avoid creating Document::quirks when InspectorDOMAgent adds an event listener
for webkitfullscreenchange event.

* Source/WebCore/dom/Node.cpp:
(WebCore::isTouchRelatedEventType):

Canonical link: https://commits.webkit.org/273722@main

aestes committed:
[tvOS] Hosting app crashes due to unrecognized selector exceptions during fullscreen media playback
https://bugs.webkit.org/show_bug.cgi?id=268331
rdar://121881715

Reviewed by Jean-Yves Avenard.

Implemented selectors expected by AVKit on WebKit's WebAVPlayerLayer, WebAVPlayerViewControllerDelegate, and WebAVMediaSelectionOption subclasses.
While here, fixed the unified sources build by adding some imports.

* Source/WebCore/platform/cocoa/WebAVPlayerLayer.mm:
(-[WebAVPlayerLayer avkit_isVisible]):
(-[WebAVPlayerLayer avkit_videoRectInWindow]):
* Source/WebCore/platform/ios/VideoPresentationInterfaceIOS.mm:
(-[WebAVPlayerViewControllerDelegate playerViewControllerShouldDismiss:]):
* Source/WebCore/platform/ios/WebAVPlayerController.mm:
(-[WebAVMediaSelectionOption copyWithZone:]):
(-[WebAVMediaSelectionOption avkit_mediaRemoteIdentifier]):
* Source/WebKit/UIProcess/Cocoa/WebPageProxyCocoa.mm:
* Source/WebKit/WebProcess/WebPage/Cocoa/WebPageCocoa.mm:

Canonical link: https://commits.webkit.org/273723@main

alanbaradlay committed:
[IFC] Containing DIV with "display: inline; white-space: nowrap;" is breaking at the gap between two contained elements
https://bugs.webkit.org/show_bug.cgi?id=268323
<rdar://121859917>

Reviewed by Antti Koivisto.

Overflowing inline box decoration (e.g. <span style="border-inline-end: solid 20px blue">) is still considered part of the inline box content
and the inline box's white-space value should apply (as opposed to its parent's style).

* LayoutTests/fast/inline/overflowing-inline-box-decoration-in-nowrap-expected.html: Added.
* LayoutTests/fast/inline/overflowing-inline-box-decoration-in-nowrap.html: Added.
* Source/WebCore/layout/formattingContexts/inline/InlineContentBreaker.cpp:
(WebCore::Layout::InlineContentBreaker::processOverflowingContent const):

Canonical link: https://commits.webkit.org/273724@main

smfr committed:
[UnifiedPDF] Can't zoom out below scale 1
https://bugs.webkit.org/show_bug.cgi?id=268310
rdar://121866624

Reviewed by Tim Horton.

For the PDF plugin, we need to allow minimum zoom scales of less than 1. Add support
for this by allowing the plugin to vend min and max scales, keeping track of them
in WebPageProxy. ViewGestureController reads them from there.

We compute the min and max elastic zoom scales relative to the min and max,
using multipliers that gives the same results as the old hard-coded values.

The PDF min and max scale values match those in PDFKit.

* Source/WebKit/UIProcess/ViewGestureController.cpp:
(WebKit::ViewGestureController::endMagnificationGesture):
* Source/WebKit/UIProcess/ViewGestureController.h:
* Source/WebKit/UIProcess/WebPageProxy.cpp:
(WebKit::WebPageProxy::minPageZoomFactor const):
(WebKit::WebPageProxy::maxPageZoomFactor const):
(WebKit::WebPageProxy::mainFramePluginHandlesPageScaleGestureDidChange):
* Source/WebKit/UIProcess/WebPageProxy.h:
* Source/WebKit/UIProcess/WebPageProxy.messages.in:
* Source/WebKit/UIProcess/mac/ViewGestureControllerMac.mm:
(WebKit::ViewGestureController::resistanceForDelta):
(WebKit::ViewGestureController::handleMagnificationGestureEvent):
(WebKit::ViewGestureController::didCollectGeometryForSmartMagnificationGesture):
* Source/WebKit/WebProcess/Plugins/PDF/PDFPluginBase.h:
(WebKit::PDFPluginBase::minScaleFactor const):
(WebKit::PDFPluginBase::maxScaleFactor const):
* Source/WebKit/WebProcess/Plugins/PDF/UnifiedPDF/PDFDocumentLayout.mm:
(WebKit::PDFDocumentLayout::layoutSingleColumn):
* Source/WebKit/WebProcess/Plugins/PluginView.cpp:
(WebKit::PluginView::initializePlugin):

Canonical link: https://commits.webkit.org/273725@main

sideshowbarker committed:
Honor “lang” attribute in no namespace only on HTML and SVG elements
https://bugs.webkit.org/show_bug.cgi?id=263655

Reviewed by Ryosuke Niwa.

https://github.com/whatwg/html/pull/9882/files added the phrase “or an
element in the SVG namespace” in the following HTML spec text:

> To determine the language of a node, user agents must use the first
appropriate step in the following list:
> …
> - If the node is an HTML element or an element in the SVG namespace,
    and it has a lang in no namespace attribute set: Use the value of
    that attribute.

And that updated spec text is now part of the HTML standard, at
https://html.spec.whatwg.org/#the-lang-and-xml:lang-attributes:svg-namespace

Gecko has already had that behavior implemented for quite a while, and
Blink was recently updated to implement it as well.

So this change fully aligns WebKit with that updated spec text, and with
the current behavior in Gecko and Blink. However, this change does that
by updating the code to *restrict* the “lang” attribute in no namespace
to *only* being honored on HTML and SVG elements.

Otherwise, without this code change, WebKit honors the “lang” attribute
in no namespace regardless of what element it’s specified on (that is,
not just for HTML elements and SVG elements) — which breaks conformance
with the spec, and breaks interoperability with Gecko and Blink.

* LayoutTests/imported/w3c/web-platform-tests/html/dom/elements/global-attributes/lang-attribute.window-expected.txt:
* Source/WebCore/dom/Element.cpp:
(WebCore::Element::attributeChanged):

Canonical link: https://commits.webkit.org/273726@main
